backend/src/schema/resolvers/searches.js

```diff
@@ -3,90 +3,200 @@ import { queryString } from './searches/queryString'
+            skip: 0,
+          },
+          countResultCallback,
+        ),
+        users: getSearchResults(context, searchUsersSetup, {
+          query: queryString(query),
+          skip: usersOffset,
+          limit: firstUsers,
+        }),
       }
-      LIMIT $limit
-      `
-
-      const userCypher = `
-      CALL db.index.fulltext.queryNodes('user_fulltext_search', $query)
-      YIELD node as resource, score
-      MATCH (resource)
-      WHERE score >= 0.0
-      AND NOT (resource.deleted = true OR resource.disabled = true)
-      RETURN resource {.*, __typename: labels(resource)[0]}
-      LIMIT $limit
-      `
-      const tagCypher = `
-      CALL db.index.fulltext.queryNodes('tag_fulltext_search', $query)
-      YIELD node as resource, score
-      MATCH (resource)
-      WHERE score >= 0.0
-      AND NOT (resource.deleted = true OR resource.disabled = true)
-      RETURN resource {.*, __typename: labels(resource)[0]}
-      LIMIT $limit
-      `
-
-      const myQuery = queryString(query)
-
-      const session = context.driver.session()
-      const searchResultPromise = session.readTransaction(async (transaction) => {
-        const postTransactionResponse = transaction.run(postCypher, {
-          query: myQuery,
-          limit,
-          thisUserId,
-        })
-        const userTransactionResponse = transaction.run(userCypher, {
-          query: myQuery,
-          limit,
-          thisUserId,
-        })
-        const tagTransactionResponse = transaction.run(tagCypher, {
-          query: myQuery,
-          limit,
-        })
-        return Promise.all([
-          postTransactionResponse,
-          userTransactionResponse,
-          tagTransactionResponse,
-        ])
-      })
-
-      try {
-        const [postResults, userResults, tagResults] = await searchResultPromise
-        log(postResults)
-        log(userResults)
-        log(tagResults)
-        return [...postResults.records, ...userResults.records, ...tagResults.records].map((r) =>
-          r.get('resource'),
-        )
-      } finally {
-        session.close()
+    },
+    searchHashtags: async (_parent, args, context, _resolveInfo) => {
+      const { query, hashtagsOffset, firstHashtags } = args
+      return {
+        hashtagCount: getSearchResults(
+          context,
+          countHashtagsSetup,
+          {
+            query: queryString(query),
+            skip: 0,
+          },
+          countResultCallback,
+        ),
+        hashtags: getSearchResults(context, searchHashtagsSetup, {
+          query: queryString(query),
+          skip: hashtagsOffset,
+          limit: firstHashtags,
+        }),
       }
     },
+    searchResults: async (_parent, args, context, _resolveInfo) => {
+      const { query, limit } = args
+      const { id: userId } = context.user
+
+      const searchType = query.replace(/^([!@#]?).*$/, '$1')
+      const searchString = query.replace(/^([!@#])/, '')
```

Comment:
> [<img alt="Tirokk" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2020-06-11T08:27:55Z" title="Thursday, June 11th 2020, 10:27:55 am +02:00">Jun 11, 2020</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -3,90 +3,200 @@ import { queryString } from './searches/queryString'
 
 // see http://lucene.apache.org/core/8_3_1/queryparser/org/apache/lucene/queryparser/classic/package-summary.html#package.description
 
+const cypherTemplate = (setup) => `
+  CALL db.index.fulltext.queryNodes('${setup.fulltextIndex}', $query)
+  YIELD node AS resource, score
+  ${setup.match}
+  ${setup.whereClause}
+  ${setup.withClause}
+  RETURN 
+  ${setup.returnClause}
+  AS result
+  SKIP $skip
+  ${setup.limit}
+`
+
+const simpleWhereClause =
+  'WHERE score >= 0.0 AND NOT (resource.deleted = true OR resource.disabled = true)'
+
+const postWhereClause = `WHERE score >= 0.0
+  AND NOT (
+    author.deleted = true OR author.disabled = true
+    OR resource.deleted = true OR resource.disabled = true
+    OR (:User {id: $userId})-[:MUTED]->(author)
+  )`
+
+const searchPostsSetup = {
+  fulltextIndex: 'post_fulltext_search',
+  match: 'MATCH (resource:Post)<-[:WROTE]-(author:User)',
+  whereClause: postWhereClause,
+  withClause: `WITH resource, author,
+  [(resource)<-[:COMMENTS]-(comment:Comment) | comment] AS comments,
+  [(resource)<-[:SHOUTED]-(user:User) | user] AS shouter`,
+  returnClause: `resource {
+    .*,
+    __typename: labels(resource)[0],
+    author: properties(author),
+    commentsCount: toString(size(comments)),
+    shoutedCount: toString(size(shouter))
+  }`,
+  limit: 'LIMIT $limit',
+}
+
+const searchUsersSetup = {
+  fulltextIndex: 'user_fulltext_search',
+  match: 'MATCH (resource:User)',
+  whereClause: simpleWhereClause,
+  withClause: '',
+  returnClause: 'resource {.*, __typename: labels(resource)[0]}',
+  limit: 'LIMIT $limit',
+}
+
+const searchHashtagsSetup = {
+  fulltextIndex: 'tag_fulltext_search',
+  match: 'MATCH (resource:Tag)',
+  whereClause: simpleWhereClause,
+  withClause: '',
+  returnClause: 'resource {.*, __typename: labels(resource)[0]}',
+  limit: 'LIMIT $limit',
+}
+
+const countSetup = {
+  returnClause: 'toString(size(collect(resource)))',
+  limit: '',
+}
+
+const countUsersSetup = {
+  ...searchUsersSetup,
+  ...countSetup,
+}
+const countPostsSetup = {
+  ...searchPostsSetup,
+  ...countSetup,
+}
+const countHashtagsSetup = {
+  ...searchHashtagsSetup,
+  ...countSetup,
+}
+
+const searchResultPromise = async (session, setup, params) => {
+  return session.readTransaction(async (transaction) => {
+    return transaction.run(cypherTemplate(setup), params)
+  })
+}
+
+const searchResultCallback = (result) => {
+  return result.records.map((r) => r.get('result'))
+}
+
+const countResultCallback = (result) => {
+  return result.records[0].get('result')
+}
+
+const getSearchResults = async (context, setup, params, resultCallback = searchResultCallback) => {
+  const session = context.driver.session()
+  try {
+    const results = await searchResultPromise(session, setup, params)
+    log(results)
+    return resultCallback(results)
+  } finally {
+    session.close()
+  }
+}
+
+const multiSearchMap = [
+  { symbol: '!', setup: searchPostsSetup, resultName: 'posts' },
+  { symbol: '@', setup: searchUsersSetup, resultName: 'users' },
+  { symbol: '#', setup: searchHashtagsSetup, resultName: 'hashtags' },
+]
+
 export default {
   Query: {
-    findResources: async (_parent, args, context, _resolveInfo) => {
-      const { query, limit } = args
-      const { id: thisUserId } = context.user
-
-      const postCypher = `
-      CALL db.index.fulltext.queryNodes('post_fulltext_search', $query)
-      YIELD node as resource, score
-      MATCH (resource)<-[:WROTE]-(author:User)
-      WHERE score >= 0.0
-      AND NOT (
-        author.deleted = true OR author.disabled = true
-        OR resource.deleted = true OR resource.disabled = true
-        OR (:User {id: $thisUserId})-[:MUTED]->(author)
-      )
-      WITH resource, author,
-      [(resource)<-[:COMMENTS]-(comment:Comment) | comment] as comments,
-      [(resource)<-[:SHOUTED]-(user:User) | user] as shouter
-      RETURN resource {
-        .*,
-        __typename: labels(resource)[0],
-        author: properties(author),
-        commentsCount: toString(size(comments)),
-        shoutedCount: toString(size(shouter))
+    searchPosts: async (_parent, args, context, _resolveInfo) => {
+      const { query, postsOffset, firstPosts } = args
+      const { id: userId } = context.user
+
+      return {
+        postCount: getSearchResults(
+          context,
+          countPostsSetup,
+          {
+            query: queryString(query),
+            skip: 0,
+            userId,
+          },
+          countResultCallback,
+        ),
+        posts: getSearchResults(context, searchPostsSetup, {
+          query: queryString(query),
+          skip: postsOffset,
+          limit: firstPosts,
+          userId,
+        }),
+      }
+    },
+    searchUsers: async (_parent, args, context, _resolveInfo) => {
+      const { query, usersOffset, firstUsers } = args
+      return {
+        userCount: getSearchResults(
+          context,
+          countUsersSetup,
+          {
+            query: queryString(query),
+            skip: 0,
+          },
+          countResultCallback,
+        ),
+        users: getSearchResults(context, searchUsersSetup, {
+          query: queryString(query),
+          skip: usersOffset,
+          limit: firstUsers,
+        }),
       }
-      LIMIT $limit
-      `
-
-      const userCypher = `
-      CALL db.index.fulltext.queryNodes('user_fulltext_search', $query)
-      YIELD node as resource, score
-      MATCH (resource)
-      WHERE score >= 0.0
-      AND NOT (resource.deleted = true OR resource.disabled = true)
-      RETURN resource {.*, __typename: labels(resource)[0]}
-      LIMIT $limit
-      `
-      const tagCypher = `
-      CALL db.index.fulltext.queryNodes('tag_fulltext_search', $query)
-      YIELD node as resource, score
-      MATCH (resource)
-      WHERE score >= 0.0
-      AND NOT (resource.deleted = true OR resource.disabled = true)
-      RETURN resource {.*, __typename: labels(resource)[0]}
-      LIMIT $limit
-      `
-
-      const myQuery = queryString(query)
-
-      const session = context.driver.session()
-      const searchResultPromise = session.readTransaction(async (transaction) => {
-        const postTransactionResponse = transaction.run(postCypher, {
-          query: myQuery,
-          limit,
-          thisUserId,
-        })
-        const userTransactionResponse = transaction.run(userCypher, {
-          query: myQuery,
-          limit,
-          thisUserId,
-        })
-        const tagTransactionResponse = transaction.run(tagCypher, {
-          query: myQuery,
-          limit,
-        })
-        return Promise.all([
-          postTransactionResponse,
-          userTransactionResponse,
-          tagTransactionResponse,
-        ])
-      })
-
-      try {
-        const [postResults, userResults, tagResults] = await searchResultPromise
-        log(postResults)
-        log(userResults)
-        log(tagResults)
-        return [...postResults.records, ...userResults.records, ...tagResults.records].map((r) =>
-          r.get('resource'),
-        )
-      } finally {
-        session.close()
+    },
+    searchHashtags: async (_parent, args, context, _resolveInfo) => {
+      const { query, hashtagsOffset, firstHashtags } = args
+      return {
+        hashtagCount: getSearchResults(
+          context,
+          countHashtagsSetup,
+          {
+            query: queryString(query),
+            skip: 0,
+          },
+          countResultCallback,
+        ),
+        hashtags: getSearchResults(context, searchHashtagsSetup, {
+          query: queryString(query),
+          skip: hashtagsOffset,
+          limit: firstHashtags,
+        }),
       }
     },
+    searchResults: async (_parent, args, context, _resolveInfo) => {
+      const { query, limit } = args
+      const { id: userId } = context.user
+
+      const searchType = query.replace(/^([!@#]?).*$/, '$1')
+      const searchString = query.replace(/^([!@#])/, '')
```


I would really like to have it commented what both regex' do in your opinion. 🤗
For me regex' are not self explaining …

Reply:
> [<img alt="Tirokk" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2020-08-27T11:15:16Z" title="Thursday, August 27th 2020, 1:15:16 pm +02:00">Aug 27, 2020</time>_

---


See my comment above ☝🏼

Reply:
> [<img alt="ogerly" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/ogerly) **Authored by [ogerly](https://github.com/ogerly)**
_<time datetime="2020-09-08T07:11:59Z" title="Tuesday, September 8th 2020, 9:11:59 am +02:00">Sep 8, 2020</time>_

---


/
^([!@#]?).*$
/
gm
^ asserts position at start of a line
1st Capturing Group ([!@#]?)
Match a single character present in the list below [!@#]?
? Quantifier — Matches between zero and one times, as many times as possible, giving back as needed (greedy)
!@# matches a single character in the list !@# (case sensitive)
.* matches any character (except for line terminators)
* Quantifier — Matches between zero and unlimited times, as many times as possible, giving back as needed (greedy)
$ asserts position at the end of a line
Global pattern flags
g modifier: global. All matches (don't return after first match)
m modifier: multi line. Causes ^ and $ to match the begin/end of each line (not only begin/end of string)